include/quic/connection.hpp

```diff
@@ -93,63 +143,146 @@ namespace oxen::quic
         Endpoint& endpoint() { return _endpoint; }
         const Endpoint& endpoint() const { return _endpoint; }
 
-      private:
-        std::shared_ptr<ContextBase> context;
-        config_t user_config;
-        Direction dir;
-        Endpoint& _endpoint;
-        const ConnectionID _source_cid;
-        ConnectionID _dest_cid;
-        Path _path;
-        std::function<void(Connection&)> on_closing;  // clear immediately after use
+        int get_streams_available() const override;
+        size_t get_max_datagram_size() const override;
+        int get_max_streams() const override { return _max_streams; }
+        bool datagrams_enabled() const override { return _datagrams_enabled; }
+        bool packet_splitting_enabled() const override { return _packet_splitting; }
+        Splitting packet_splitting_policy() const override { return _policy; }
 
+      private:
         // private Constructor (publicly construct via `make_conn` instead, so that we can properly
         // set up the shared_from_this shenanigans).
         Connection(
                 Endpoint& ep,
                 const ConnectionID& scid,
                 const ConnectionID& dcid,
                 const Path& path,
-                std::shared_ptr<ContextBase> ctx,
-                Direction dir,
+                std::shared_ptr<IOContext> ctx,
                 ngtcp2_pkt_hd* hdr = nullptr);
 
+        ngtcp2_crypto_conn_ref cref;
+        Endpoint& _endpoint;
+        std::shared_ptr<IOContext> context;
+        Direction dir;
+        const ConnectionID _source_cid;
+        ConnectionID _dest_cid;
+        Path _path;
+        std::function<void(Connection&)> on_closing;  // clear immediately after use
+        const int _max_streams{DEFAULT_MAX_BIDI_STREAMS};
+        const bool _datagrams_enabled{false};
+        const bool _packet_splitting{false};
+        const Splitting _policy{Splitting::NONE};
+        datagram_recv_callback dgram_data_cb;
+
+        /// Datagram Numbering:
```

Comment:
I like how this is laid out.  I'd say that `packet` and `datagram` could be used more precisely here though, to avoid confusion.  For example, I'd say it should be phrased that we might split "packets" into multiple (two) datagrams.  Trying to cover all cases:

-----

> Each datagram ID is incremented by four from the previous one, regardless of whether we are splitting ~~datagrams~~ **packets**.

> In the unsplit packet scheme, the dgram ID of each ~~packet~~ **datagram** satisfies the rule:

> In the split-packet scheme, the dgram ID of the first of two ~~packets~~ **datagrams** satisfies the rule:

> The second of the two ~~packets~~ datagrams satisfies the rule:

> As a result, a packet-splitting endpoint should never send or receive a ~~packet~~ datagram whose ID is a perfect multiple of 4

Reply:
Thank you for the edits! That all makes perfect sense, I'll make the change